sematic/external_resource.py

```diff
@@ -0,0 +1,331 @@
+        #   - possibly an error with the resource. Status message should have more info.
+        ResourceState.DEACTIVATING,
+    },
+    ResourceState.DEACTIVATING: {
+        # Deactivating -> Deactivated: normal progression for deactivation
+        ResourceState.DEACTIVATED,
+    },
+    ResourceState.DEACTIVATED: {},
+}
+
+
+@dataclass(frozen=True)
+class ResourceStatus:
+    """The status of the resource.
+
+    Attributes
+    ----------
+    state:
+        The discrete state that the resource is in
+    message:
+        A human-readable message about why the resource entered this state
+    last_update_epoch_time:
+        The last time this status was checked against the actual external resource
+    """
+
+    state: ResourceState
+    message: str
+    last_update_epoch_time: int = field(default_factory=lambda: int(time.time()))
+
+
+T = TypeVar("T")
+
+
+@dataclass(frozen=True)
+class ExternalResource:
+    """Represents a resource tracked by Sematic for usage in Sematic funcs.
+
+    Examples of possible external resources include small data processing
+    clusters or distributed training clusters.
+
+    Specific external resources should inherit from this class. They may add their
+    own dataclass attributes and must implement the _do_activate, _do_deactivate, and
+    _do_update methods on this class.
+
+    This class represents immutable objects. As with immutable strings, the
+    preferred pattern is to replace instances with updated copies rather
+    than modify in-place. This allows for better control over validation of
+    objects and state transitions.
+
+    Attributes
+    ----------
+    id:
+        A UUID representing this particular resource instance
+    status:
+        Information about the current status of the resource
+    """
+
+    id: str = field(default_factory=lambda: uuid.uuid4().hex)
+    status: ResourceStatus = field(
+        default_factory=lambda: ResourceStatus(
+            state=ResourceState.CREATED,
+            message="Resource has not been activated yet",
+        )
+    )
+
+    def __post_init__(self):
+        for field_ in fields(self):
+            field_value = getattr(self, field_.name)
+            if isinstance(field_value, AbstractFuture):
+                func_name = {field_value.calculator.__name__}
+                type_name = type(self).__name__
+                raise IllegalUseOfFutureError(
+                    f"Tried to instantiate {type_name} with a future for the "
+                    f"value of the field '{field_.name}'. The future appears to be the "
+                    f"return value from a call to '{func_name}'. "
+                    f"If you wish to use {type_name} downstream from "
+                    f"'{func_name}', consider wrapping the code "
+                    f"using {type_name} in a new Sematic func and passing the "
+                    f"output of {func_name} to that new func."
+                )
```

Comment:
Is this likely to happen? Does the `SparkResource` class expect any init fields?